servlet/src/main/java/io/undertow/servlet/spec/ServletPrintWriter.java

```diff
@@ -254,45 +236,86 @@ public void write(final char[] buf, final int off, final int len) {
         write(cb);
     }
 
+    private static int writeAndFlushAscii(ServletOutputStreamImpl outputStream, ByteBuffer buffer, char[] chars, int start, int end) throws IOException {
+        assert buffer.order() == ByteOrder.BIG_ENDIAN;
+        int i = start;
+        while (i < end) {
+            final int bufferPos = buffer.position();
+            final int bufferRemaining = buffer.remaining();
+            final int sRemaining = end - i;
+            final int remaining = Math.min(sRemaining, bufferRemaining);
+            final int written = setAsciiBE(buffer, bufferPos, chars, i, remaining);
+            i += written;
+            buffer.position(bufferPos + written);
+            if (!buffer.hasRemaining()) {
+                outputStream.flushInternal();
+            }
+            // we have given up with the fast path? slow path NOW!
+            if (written < remaining) {
+                return i;
+            }
+        }
+        return i;
+    }
+
+    private static int setAsciiBE(ByteBuffer buffer, int out, char[] chars, int off, int len) {
+        final int longRounds = len >>> 3;
+        for (int i = 0; i < longRounds; i++) {
+            final long batch1 = (((long) chars[off]) << 48) |
```

Comment:
@theRealAph not very happy about these casts; I'm going to refresh the knowledge of JLS, but without them (or `& 0xFFFFL`)  the chars read were treated as `int` and too large shifts were just truncating/losing data

Reply:
Cast looks correct to me.

Reply:
Nice, thanks!